### ISSUE TITLE
Peons processes left running detached from main Middle Manager process

### ISSUE THREAD

**Issue by spektom** [implydata/distribution#6]
Not sure what happened, but there are processes on all data nodes, which are not attached to the parent middle manager process anymore. For example, look at the last two processes in this process tree:

```
  2270 ?        Ss     2:09 /usr/bin/python /usr/bin/supervisord -c /etc/supervisor/supervisord.conf
 23780 ?        S<    28:13  \_ /usr/bin/af-sniffer eth0 /etc/af-sniffer.cfg
 75320 ?        S      0:28  \_ perl /opt/imply/bin/supervise -c /opt/imply/conf/supervise/data.conf
 75368 ?        Ssl  902:59      \_ java -server -Xms12g -Xmx12g -XX:NewSize=6g -XX:MaxNewSize=6g -XX:MaxDirectMemorySize=32g -XX:+UseConcMarkSweepGC -XX:+PrintGCDetails -XX
 75369 ?        Ssl    7:36      \_ java -server -Xms64m -Xmx64m -Duser.timezone=UTC -Dfile.encoding=UTF-8 -Djava.io.tmpdir=/mnt/druid-tmp -Djava.util.logging.manager=org.ap
 27200 ?        Sl    18:19          \_ java -cp /opt/imply-1.1.0/conf/druid/_common:/opt/imply-1.1.0/conf/druid/middleManager:/opt/imply-1.1.0/bin/../dist/druid/lib/Roaring
113962 ?        Sl    15:41          \_ java -cp /opt/imply-1.1.0/conf/druid/_common:/opt/imply-1.1.0/conf/druid/middleManager:/opt/imply-1.1.0/bin/../dist/druid/lib/Roaring
119100 ?        Sl     9:36          \_ java -cp /opt/imply-1.1.0/conf/druid/_common:/opt/imply-1.1.0/conf/druid/middleManager:/opt/imply-1.1.0/bin/../dist/druid/lib/Roaring
119161 ?        Sl    11:19          \_ java -cp /opt/imply-1.1.0/conf/druid/_common:/opt/imply-1.1.0/conf/druid/middleManager:/opt/imply-1.1.0/bin/../dist/druid/lib/Roaring
  2357 ?        Ss     0:00 /sbin/mdadm --monitor --pid-file /run/mdadm/monitor.pid --daemonise --scan --syslog
  2467 ?        Ss     0:51 /usr/sbin/ntpd -p /var/run/ntpd.pid -g -u 106:111
  2525 tty1     Ss+    0:00 /sbin/getty -8 38400 tty1
  2526 ttyS0    Ss+    0:00 /sbin/getty -8 38400 ttyS0
 21591 ?        Ssl    0:44 /usr/sbin/nscd
 22215 ?        Sl    90:41 consul agent -client=10.10.8.53 -dc=eu1 -join consulmaster -pid-file=/var/run/consul.pid -data-dir /opt/consul -config-dir /etc/consul.d -bind 10
 23667 ?        Ss     0:00 /usr/sbin/nrsysmond -c /etc/newrelic/nrsysmond.cfg -p /var/run/newrelic/nrsysmond.pid
116290 ?        Sl    20:01  \_ /usr/sbin/nrsysmond -c /etc/newrelic/nrsysmond.cfg -p /var/run/newrelic/nrsysmond.pid
 23700 ?        Ss     0:00 /usr/sbin/collectdmon -P /var/run/collectd.pid -- -C /etc/collectd/collectd.conf
 62790 ?        Sl    13:30  \_ collectd -C /etc/collectd/collectd.conf -f
 23767 ?        Sl     8:45 /opt/sensu/embedded/bin/ruby /opt/sensu/bin/sensu-client -b -c /etc/sensu/config.json -d /etc/sensu/conf.d -e /etc/sensu/extensions -p /var/run/s
 83754 ?        Sl   111:46 java -cp /opt/imply-1.1.0/conf/druid/_common:/opt/imply-1.1.0/conf/druid/middleManager:/opt/imply-1.1.0/bin/../dist/druid/lib/RoaringBitmap-0.5.1
 96154 ?        Sl    66:50 java -cp /opt/imply-1.1.0/conf/druid/_common:/opt/imply-1.1.0/conf/druid/middleManager:/opt/imply-1.1.0/bin/../dist/druid/lib/RoaringBitmap-0.5.1

```

**Comment by gianm** [comment]
Hey @spektom,

Can you tell if these are doing anything? If you `jstack -l 83754` or maybe `sudo -u <your imply user> jstack -l 83754` do you see anything interesting in the stack traces?

There's some code in the peons that should cause them to self-terminate when they detect that their parent has disappeared. So if they're not going away, that makes me think they're hung somewhere.

**Comment by spektom** [comment]
Hi @gianm,

Looks like there's deadlock, at least it's reported by `jstack`:

```
root@druiddata-20001-001-prod:~# jstack -l 83754
83754: Unable to open socket file: target process not responding or HotSpot VM not loaded
The -F option can be used when the target process is not responding
root@druiddata-20001-001-prod:~# jstack -F -l 83754
Attaching to process ID 83754, please wait...
Debugger attached successfully.
Server compiler detected.
JVM version is 25.72-b15
Deadlock Detection:

Found one Java-level deadlock:
=============================

"inapps-incremental-persist":
  waiting to lock Monitor@0x00007f005400b938 (Object@0x00000006c445ac00, a java/util/HashMap),
  which is held by "Thread-27"
"Thread-27":
  waiting to lock Monitor@0x00007f004404da98 (Object@0x00000006c529b920, a io/druid/guice/LifecycleScope$1),
  which is held by "inapps-incremental-persist"

Found a total of 1 deadlock.
```

**Comment by spektom** [comment]
My workaround currently is running the following script periodically:

```bash
#!/bin/bash

for pid in $(jps -m | grep "peon" | awk '{print $1}'); do
  if [ $(ps -o ppid= -p $pid) -eq 1 ]; then
    echo "Killing detached peon (PID: $pid)"
    kill -9 $pid
  fi
done
```

**Comment by gianm** [comment]
@spektom, we fixed a task-lifecycle-related deadlock in https://github.com/druid-io/druid/pull/2581, although I am not sure if this is what you're seeing or not. It definitely looks similar enough to possibly be the same thing, but different enough to possibly be a different thing.

At any rate- that deadlock fix will be in the next Imply and Druid releases, so it may help you too.

We can confirm if the issue you're seeing is the same as the one that was fixed if you could get a fuller thread dump. I think the way to do this would be to try to catch one of the task peons in this stuck state _before_ the supervising middleManager is restarted (you should be able to find one by looking at your overlord console for tasks that should have exited a while ago, but didn't). At that point you should be able to `jstack -l` it and get stack traces of all the threads.

**Comment by gianm** [comment]
Hey @spektom, I'm going to close this under the assumption that this is related to the deadlock issue. This has been fixed in IAP 1.1.1, so please let us know if you see it there.
